### ISSUE TITLE
Allow something other than the field "name" property to be the default when showing error messages.

### ISSUE THREAD

**Issue by evantrimboli** [ant-design/ant-design#9954]
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
By default, when an error message is displayed is uses `"%s is required"`, where `%s` is the name of the field. The field name isn't necessarily something to present to the user. For example, my field name might be `parentId`, but to the user I want to display `Parent`. In essence, across the entirety of my app I want to use the label as opposed to the name to display errors.

### What does the proposed API look like?
I had a browse around the source, it seems as though `fullField` is an attempt to allow this, however it doesn't appear to work correctly. See: https://codesandbox.io/s/v06vq5zlx0 which attempts to use fullField.

A potential API would be something like:

```
rules: [{
  required: true,
  errorProp: 'Parent'
}]
```

**Comment by dengfuping** [comment]
@evantrimboli It's unnecessary to supply such a feature. You could customize your error message using `message` property.

**Comment by evantrimboli** [comment]
Of course I could customize the label, but that's not the point. There is no ability to have a convention based naming system.

For example, all the field names in my app are lower case. This means whenever an error is displayed, it will show as `name is required`. It would make more sense for it to read `Name is required`.

Now say I have 10 fields in a form. For each of those required field, I need to manually set the message because I want the required message to use `Name` and not `name`. This requires a bunch of repetition where it could be easily solved using a transform function.

There should a way, either globally or potentially per rule some way to automatically transform the the field name before it is sent to the formatter.

**Comment by SyuTingSong** [comment]
@evantrimboli Similar problem I hit. However, the `fullField` in rules is not the label or display name or such things, `fullField` is the full pathname for depp object validating. I create a feature request issue to the underlying lib [async-validator](https://github.com/yiminghe/async-validator/issues/97) project for a new `localField` for message displaying localization field name.

**Comment by evantrimboli** [comment]
Great, thanks for your post! It would be nice to be able to:
a) Specify a particular value for a field (basically the label) to also be used for the error
b) Have some kind of transform function that could be applied across your app, for example:

`const transform = s => s.split('_').map(capitalizeFn).join(' ')`

**Comment by rohit-smpx** [comment]
Any way to do this now? I'd like to use the `label` property for validator error messages

**Comment by csandanov** [comment]
I was also surprised to find out it that names are used instead of labels in error messages. I believe it should be label by default with ability to customize because a label is what a user sees, not `name` property. Especially this is a problem when you have multiple fields and you have to make them unique, e.g. `name-${id}`. 

I tried to customize it via:
```js
const validateMessages = {
  required: "'${label}' is required!",
};
```
but it doesn't recognize the label property for some reason.

**Comment by rohit-smpx** [comment]
> I tried to customize it via:
> 
> ```js
> const validateMessages = {
>   required: "'${label}' is required!",
> };
> ```
> 
> but it doesn't recognize the label property for some reason.

The message is made by the rule validator and not the form so it doesn't have access to `label`. 

If you send the label in the rule, like : `{required: true, label: 'Email ID'}` it will show properly but I don't think that is officially supported.

**Comment by csandanov** [comment]
@rohit-gohri yes, that works, although I'm using typescript and gives a warning that `label` is not assignable to `Rule[]`

**Comment by devuxer** [comment]
@rohit-gohri & @csandanov,

@rohit-gohri 's suggestion is actually pretty promising! Perhaps it could just be made official and added to the documentation.

I was even able to get it to work with TypeScript with these three steps:

1. Copy the default validation messages from [the source](https://github.com/react-component/field-form/blob/master/src/utils/messages.ts).
2. Replace `${name}` with `${label}`.
3. Add this simple utility function:
    ```ts
    // add label property to each rule
    export function getRules(label: string, ...rules: Rule[]): Array<Rule & { label: string }> {
        return rules.map(x => ({ ...x, label }));
    }
    ```

Usage for required + len:

```tsx
<Form.Item name="zipCode" noStyle rules={getRules("Zip Code", { required: true, len: 5 })}>
    <Input placeholder="Zip Code" />
</Form.Item>
```

Usage for required + pattern w/ custom message:

```tsx
<Form.Item
    name="legalName"
    label="Legal Name"
    extra="Full, official name {e.g., Robert Smith)"
    rules={getRules(
        "Legal Name",
        { required: true },
        {
            pattern: /^ *[A-Za-z-',.]+( +[A-Za-z-',.]+)+$ */,
            message: `'Legal Name' must have at least two words and only these symbols: - ' , . `
        }
    )}
>
    <Input />
</Form.Item>
```

**Comment by rohit-gohri** [comment]
@devuxer I don't think you even need to do that now, label support was added in the 4.1 release : https://github.com/ant-design/ant-design/pull/21835
Though the rules hack is still useful if you want to use it for some other custom prop.

**Comment by devuxer** [comment]
@rohit-gohri,

Ahh, I didn't realize. Thanks for letting me know 👍

**Comment by yoyo837** [comment]
Can this be closed?

**Comment by afc163** [comment]
Reviewed against the latest `master` (`0b1021909f`). Form validation templates now support the display label separately from the field name.

Label message variables were added in v4.1 by #21835 and remain in the latest Form implementation.

Closing this issue accordingly. If a minimal reproduction still demonstrates different behavior on the current `master`, please open a new issue with that reproduction.
